src/ai/backend/manager/models/image.py

```diff
@@ -142,44 +143,74 @@ async def rescan_images(
+        all_registry_config = cast(
             dict[str, ContainerRegistryRow],
-            {row.registry_name: row for row in result.scalars().all()},
+            {
+                f"{registry_row.registry_name}/{registry_row.project}": registry_row
+                for registry_row in result.scalars().all()
+            },
         )
 
     # TODO: delete images from registries removed from the previous config?
     if registry_or_image is None:
         # scan all configured registries
-        registries = latest_registry_config
+        registries = all_registry_config
     else:
         # find if it's a full image ref of one of configured registries
-        for registry_name, registry_info in latest_registry_config.items():
-            if registry_or_image.startswith(registry_name + "/"):
-                repo_with_tag = registry_or_image.removeprefix(registry_name + "/")
-                log.debug(
-                    "running a per-image metadata scan: {}, {}",
-                    registry_name,
-                    repo_with_tag,
+        matching_registries = filter_registry_dict(
+            all_registry_config,
+            lambda registry_key, _row: registry_or_image.startswith(registry_key + "/"),
+        )
+
+        if matching_registries:
+            if len(matching_registries) > 1:
+                raise RuntimeError(
+                    "ContainerRegistryRows exist with the same registry_name and project!",
                 )
-                scanner_cls = get_container_registry_cls(registry_info)
-                scanner = scanner_cls(db, registry_name, registry_info)
-                await scanner.scan_single_ref(repo_with_tag)
-                return
+
+            registry_key, registry_row = next(iter(matching_registries.items()))
+            registry_name = ImageRef.parse_image_str(registry_key, "*").registry
+            repo_with_tag = registry_or_image.removeprefix(registry_name + "/")
+            log.debug("running a per-image metadata scan: {}, {}", registry_name, repo_with_tag)
+
+            scanner_cls = get_container_registry_cls(registry_row)
+            scanner = scanner_cls(db, registry_name, registry_row)
+            await scanner.scan_single_ref(repo_with_tag)
+            return
         else:
             # treat it as a normal registry name
+            log.debug("running a per-registry metadata scan")
+
             registry = registry_or_image
-            try:
-                registries = {registry: latest_registry_config[registry]}
-                log.debug("running a per-registry metadata scan")
-            except KeyError:
+
+            registries = filter_registry_dict(
+                all_registry_config,
+                lambda registry_key, _row: ImageRef.parse_image_str(registry_key, "*").registry
+                == registry,
+            )
+
+            if not registries:
                 raise RuntimeError("It is an unknown registry.", registry)
```

Comment:
Wouldn’t it be possible to refactor this part into a separate function that returns the registries?